File: elixir/word-count/word-count.exs
defmodule Words do

  def count(sentence) do
    normalize_sentence(sentence) |> sentence_to_words |> list_to_counted_hash
  end

  defp normalize_sentence(sentence) do
    String.downcase(sentence) 
  end

  defp sentence_to_words(sentence) do
    Regex.scan(%r/\w+/, sentence) |> List.flatten
  end

  defp list_to_counted_hash(list) do
    Enum.reduce list, HashDict.new, fn(word, dict) ->
      HashDict.update(dict, word, 1, fn(val) -> val + 1 end)
    end
  end

end

File: clojure/word-count/word_count.clj
(ns phrase
  (:require [clojure.string :as string]))

(defn- normalize-sentence [sentence] 
  (string/lower-case sentence))

(defn- sentence-to-words [sentence]
  (re-seq #"\w+" sentence))

(defn- words-to-counted-hash [words]
  (frequencies words))

(defn word-count [sentence]
  (-> sentence normalize-sentence
               sentence-to-words
               words-to-counted-hash))
